controllers/component_controller.go

```diff
@@ -47,16 +55,105 @@ type ComponentReconciler struct {
+		logger.Info("Component is being deleted")
+
+		// Remove finalizer to complete the deletion.
+		instance.Finalizers = utils.RemoveString(instance.Finalizers, corev1alpha1.ComponentFinalizer)
+		if err := r.Client.Update(ctx, instance); err != nil {
+			logger.Error(err, "Failed to remove component finalizer")
+		}
+		return reconcile.Result{}, err
+	}
+
+	done, err := r.UpdateComponent(ctx, logger, instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	} else if !done {
+		return reconcile.Result{}, nil
+	}
+
+	logger.Info("Synchronized component successfully")
 	return ctrl.Result{}, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ComponentReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&corev1alpha1.Component{}).
+		For(&corev1alpha1.Component{}, builder.WithPredicates(predicate.Funcs{
+			UpdateFunc: r.OnComponentUpdate,
+		})).
 		Complete(r)
 }
+
+// UpdateComponent updates new component, add finalizer if necessary.
+func (r *ComponentReconciler) UpdateComponent(ctx context.Context, logger logr.Logger, instance *corev1alpha1.Component) (bool, error) {
+	done := true
+	// check finalizer, add finalizer and report not done (need another update event)
+	l := len(instance.Finalizers)
+	instance.Finalizers = utils.AddString(instance.Finalizers, corev1alpha1.ComponentFinalizer)
+	if l != len(instance.Finalizers) {
+		done = false
+		logger.V(4).Info("Component finalizer added", "Finalizer", corev1alpha1.ComponentFinalizer)
+		err := r.Client.Update(ctx, instance)
+		if err != nil {
+			logger.Error(err, "Failed to add component finalizer")
+		}
+		return done, err
+	}
+
+	// check if ownerReferences exist, report done (nothing to do)
+	if len(instance.OwnerReferences) == 0 {
+		return done, nil
+	}
+
+	var repoName string
+	// find the name of the repository
+	for _, n := range instance.OwnerReferences {
+		if n.Kind == "Repository" {
+			repoName = n.Name
+			break
+		}
+	}
+
+	// check if certain tag exist, report not done (need another update event)
+	if v, ok := instance.Labels[corev1alpha1.ComponentRepositoryLabel]; !ok || v != repoName {
+		// add component.repository=<repository-name> to labels
+		done = false
+		instance.Labels[corev1alpha1.ComponentRepositoryLabel] = repoName
+		logger.V(4).Info("Component repository label added", "Label", corev1alpha1.ComponentRepositoryLabel)
+		err := r.Client.Update(ctx, instance)
+		if err != nil {
+			logger.Error(err, "Failed to add component repository label")
+		}
+		return done, err
+	}
+
+	return done, nil
+}
+
+// OnComponentUpdate checks if a reconcile process is needed when updating. Default true.
+func (r *ComponentReconciler) OnComponentUpdate(event event.UpdateEvent) bool {
+	old := event.ObjectOld.(*corev1alpha1.Component)
+	new := event.ObjectNew.(*corev1alpha1.Component)
+
+	return old.ResourceVersion != new.ResourceVersion || !reflect.DeepEqual(old.Status, new.Status)
```

Comment:
@Abirdcfly Any suggestion? `ResourceVersion` will also change every time the resource changes, so is the judgment here redundant?

Reply:
I have seen some **rare** cases in codes in case the `ResourceVersion` value was compared. I have an open opinion on this part of the code and this comparison can be kept or removed.